### PR TITLE
MSHARED-1039: Fix array parsing

### PR DESCRIPTION
**Pull request by chonton**
Arrays [of Arrays]+ of class types should add just the class type to dependency set

Arrays of scalar types should not be added

 - [ x] I hereby declare this contribution to be licenced under the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0)

 - [ x] In any other case, please file an [Apache Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

### CODE REVIEW

**Comment by slawekjaranowski** [comment]
Thanks. Will be good to add some of test. Unit or IT, of course we can extends existing one.

File: src/test/java/org/apache/maven/shared/dependency/analyzer/asm/ResultCollectorTest.java [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+import java.io.InputStream;
+import java.util.Set;
+
+import junit.framework.TestCase;
+import org.apache.maven.shared.dependency.analyzer.testcases.ArrayCases;
+import static org.assertj.core.api.Assertions.assertThat;
+import org.junit.Test;
+
+public class ResultCollectorTest
+    extends TestCase
```
Comment (slawekjaranowski) [inline src/test/java/org/apache/maven/shared/dependency/analyzer/asm/ResultCollectorTest.java:32]: extends not needed for junit4

File: src/test/java/org/apache/maven/shared/dependency/analyzer/asm/ResultCollectorTest.java [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+    @Test
+    public void testArrayCases()
+        throws IOException
+    {
+        Set<String> dependencies = getDependencies( ArrayCases.class );
+        assertThat( dependencies ).doesNotContain( "[I" );
+        for ( String dependency : dependencies )
+        {
+            assertThat( dependency ).doesNotStartWith( "[" );
+        }
```
Comment (slawekjaranowski) [inline src/test/java/org/apache/maven/shared/dependency/analyzer/asm/ResultCollectorTest.java:56]: We can use:

```
assertThat( dependencies ).allSatisfy( dependency -> assertThat( dependency ).doesNotStartWith( "[" ) );
```

File: src/main/java/org/apache/maven/shared/dependency/analyzer/asm/ResultCollector.java [hunk 3]
```diff
@@ -57,15 +57,24 @@ public void addName( String name )
...
-        if ( name.startsWith( "[L" ) && name.endsWith( ";" ) )
+        if ( name.charAt( 0 ) == '[' )
         {
-            name = name.substring( 2, name.length() - 1 );
+            int i = 0;
+            do
+            {
+                ++i;
+            }
+            while ( name.charAt( i ) == '[' ); // could have array of array ...
```
Comment (slawekjaranowski) [inline src/main/java/org/apache/maven/shared/dependency/analyzer/asm/ResultCollector.java:67]: can `name` contains multiple `[` ... ?
we don't check length of name - can couse `IndexOutOfBoundsException`

**Review by slawekjaranowski — CHANGES_REQUESTED** [review]

File: src/main/java/org/apache/maven/shared/dependency/analyzer/asm/ResultCollector.java (hunk 3 above)
Comment (chonton) [reply src/main/java/org/apache/maven/shared/dependency/analyzer/asm/ResultCollector.java:67]: name can contain any number of '[' - see [Field Descriptors](https://docs.oracle.com/javase/specs/jvms/se17/html/jvms-4.html#jvms-4.3.2) in jvm spec.
Yes, we could have IndexOutOfBoundsException if class is malformed.  Does it matter what sort of exception is thrown?

File: src/test/java/org/apache/maven/shared/dependency/analyzer/asm/ResultCollectorTest.java (hunk 1 above)
Comment (chonton) [reply src/test/java/org/apache/maven/shared/dependency/analyzer/asm/ResultCollectorTest.java]: done

File: src/test/java/org/apache/maven/shared/dependency/analyzer/asm/ResultCollectorTest.java (hunk 2 above)
Comment (chonton) [reply src/test/java/org/apache/maven/shared/dependency/analyzer/asm/ResultCollectorTest.java]: done

File: src/main/java/org/apache/maven/shared/dependency/analyzer/asm/ResultCollector.java (hunk 3 above)
Comment (chonton) [reply src/main/java/org/apache/maven/shared/dependency/analyzer/asm/ResultCollector.java:67]: @slawekjaranowski - any thoughts?

**Review by slawekjaranowski — APPROVED** [review]

**Pull request merged**